exercises/chapter8/test/Main.purs

```diff
@@ -1,11 +1,136 @@
 module Test.Main where
 
 import Prelude
-
+import Control.Monad.Free (Free)
+import Data.List
+  ( List(..)
+  , foldM
+  , singleton
+  , (:)
+  )
+import Data.Maybe (Maybe(..))
 import Effect (Effect)
-import Effect.Class.Console (log)
+{- | Enable your own code to be tested by removing this comment
+import MySolutions
+  ( third
+  )
+-}
+{- | Comment out this import to instead apply the unit tests to your MySolutions above -}
+import Official.Solutions
+  ( filterM
+  , possibleSums
+  , third
+  )
+{--}
+import Test.Examples (countThrows, safeDivide)
+import Test.Unit (TestF, suite, test)
+import Test.Unit.Assert as Assert
+import Test.Unit.Main (runTest)
 
 main :: Effect Unit
-main = do
-  log "🍝"
-  log "You should add some tests."
+main =
+  runTest do
+    runChapterExamples
+    suite "Verifying unit test environment" do
+      test "Initial passing test"
+        $ Assert.equal true true
+    suite "Exercises Group 1" do
+      suite "Exercise 1 - function third" do
+        test "No elements"
+          $ Assert.equal Nothing
+          $ third ([] :: Array Int)
+        test "1 element"
+          $ Assert.equal Nothing
+          $ third [ 1 ]
+        test "2 elements"
+          $ Assert.equal Nothing
+          $ third [ 1, 2 ]
+        test "3 elements"
+          $ Assert.equal (Just 3)
+          $ third [ 1, 2, 3 ]
+        test "4 elements"
+          $ Assert.equal (Just 4)
+          $ third [ 1, 2, 4, 3 ]
+      suite "Exercise 2 - function possibleSums" do
+        test "[]"
+          $ Assert.equal [ 0 ]
+          $ possibleSums []
+        test "[1, 2, 10]"
+          $ Assert.equal [ 0, 1, 2, 3, 10, 11, 12, 13 ]
+          $ possibleSums [ 1, 2, 10 ]
+      suite "Exercise 3 - ap and apply" do
+        test "NO TESTS"
+          $ Assert.equal true true
+      suite "Exercise 4 - Monad laws for Maybe" do
+        test "NO TESTS"
+          $ Assert.equal true true
+      suite "Exercise 5 - function filterM" do
+        {- | Couldn't get Array monad to work:
+             1.  There is no filterM for Data.Array
+             2.  List.filterM expects the function return result to be List, not Array
+
+             What am I doing wrong?
+        -}
```

Comment:
I don't think you're doing anything wrong. This passage caused me a bit of confusion on first read:

> 1. (Medium) Write a function `filterM` which generalizes the `filter` function on lists. Your function should have the following type signature:
>
>     ```haskell
>     filterM :: forall m a. Monad m => (a -> m Boolean) -> List a -> m (List a)
>     ```
>
>     Test your function in PSCi using the `Maybe` and `Array` monads.

I'm not sure if your confusion is the same as mine, but what I _think_ is meant here, after a few re-reads, is not to implement `filterM` by replacing `List` with `Array` and `m` with `Maybe`, but to check whether both 
```filterM :: forall a. (a -> Array Boolean) -> List a -> Array (List a)```
and
```filterM :: forall a. (a -> Maybe Boolean) -> List a -> Maybe (List a)```
work as expected.

The behavior (at least for `m ~ Array`) is not immediately intuitive (at least for me 😄 ) but [it seems to be a well-known trick among Haskellers](https://www.google.com/search?q=haskell+filterM+power+set) that `filterM (\_ -> [True, False]) xs` produces a power set of `xs`.

Looks like you've got the `Maybe` test suite, I think the relevant tests using `Array` could be
```
        suite "Array Monad" do
          let
            pairOfBooleans :: forall a. a -> Array Boolean
            pairOfBooleans = \_ -> [true, false]
          test "Power set"
            $ Assert.equal 
              [ (1 : 2 : 3 : Nil)
              , (1 : 2 : Nil)
              , (1 : 3 : Nil)
              , (1 : Nil)
              , (2 : 3 : Nil)
              , (2 : Nil)
              , (3 : Nil)
              , Nil ]
            $ filterM 
              pairOfBooleans
              ( 1 : 2 : 3 : Nil)
```
(Disclaimer: I'm a newb with PS and might be mistaken 🤷 )